c++/src/capnp/compat/byte-stream.c++

```diff
@@ -795,20 +805,61 @@ public:
         }
       }
       KJ_CASE_ONEOF(capnpStream, capnp::ByteStream::Client*) {
-        // TODO(perf): Split very large writes into many smaller writes.
+        auto writePieces = [&capnpStream](kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) {
+          size_t size = 0;
+          for (auto& piece: pieces) size += piece.size();
+          auto req = capnpStream->writeRequest(MessageSize { 8 + size / sizeof(word), 0 });
+          auto out = req.initBytes(size);
+          byte* ptr = out.begin();
+          for (auto& piece: pieces) {
+            memcpy(ptr, piece.begin(), piece.size());
+            ptr += piece.size();
+          }
+          KJ_ASSERT(ptr == out.end());
+          return req.send();
+        };
+
+        static const uint MAX_BYTES_PER_WRITE = 1 << 16;
         size_t size = 0;
         for (auto& piece: pieces) size += piece.size();
-        auto req = capnpStream->writeRequest(MessageSize { 8 + size / sizeof(word), 0 });
+        if (size <= MAX_BYTES_PER_WRITE) {
+          return writePieces(pieces);
+        } else {
+          // ughhhhhhhhhh, we need to split the pieces.
```

Comment:
Since this code is nearly identical to above, and indeed is doing pretty much the same thing, could we factor out a lambda or something?

Reply:
Or a helper method?

Reply:
I put up a fixup, I'm not sure if it's what you had in mind.

But I'm really confused/stuck by why I'm getting a complaint about the deleted copy constructor in my `attach` of any `Array` here... https://github.com/capnproto/capnproto/blob/fd4fea1b37e5ffb82d6fb61be559c47505405969/c%2B%2B/src/capnp/compat/byte-stream.c%2B%2B#L774

```
    In file included from /ekam-provider/canonical/capnp/compat/byte-stream.c++:22:
    In file included from /ekam-provider/canonical/capnp/compat/byte-stream.h:25:
    In file included from /ekam-provider/c++header/capnp/compat/byte-stream.capnp.h:9:
    In file included from /ekam-provider/c++header/capnp/capability.h:28:
    In file included from /ekam-provider/c++header/kj/async.h:24:
    In file included from /ekam-provider/c++header/kj/async-prelude.h:28:
    /ekam-provider/c++header/kj/tuple.h:314:12: error: call to deleted constructor of 'Decay<const kj
      ::Array<kj::ArrayPtr<const unsigned char>>>' (aka 'kj::Array<kj::ArrayPtr<const unsigned char>>
      ')
    /ekam-provider/c++header/kj/tuple.h:257:12: note: in instantiation of function template 
      specialization 'kj::_::MakeTupleFunc::operator()<const kj::Array<kj::ArrayPtr<const unsigned 
      char>>>' requested here
    /ekam-provider/c++header/kj/tuple.h:245:10: note: in instantiation of function template 
      specialization 'kj::_::ExpandAndApplyFunc<kj::_::MakeTupleFunc, const kj::Array<kj::ArrayPtr<
      const unsigned char>>>::operator()<>' requested here
    /ekam-provider/c++header/kj/tuple.h:265:10: note: in instantiation of function template 
      specialization 'kj::_::expandAndApply<kj::_::ExpandAndApplyFunc<kj::_::MakeTupleFunc, const kj:
      :Array<kj::ArrayPtr<const unsigned char>>>>' requested here
    /ekam-provider/c++header/kj/tuple.h:348:13: note: in instantiation of function template 
      specialization 'kj::_::expandAndApply<kj::_::MakeTupleFunc, const kj::Array<kj::ArrayPtr<const 
      unsigned char>>>' requested here
    /ekam-provider/c++header/kj/async-inl.h:1040:25: note: in instantiation of function template 
      specialization 'kj::tuple<const kj::Array<kj::ArrayPtr<const unsigned char>>>' requested here
    /ekam-provider/canonical/capnp/compat/byte-stream.c++:774:32: note: in instantiation of function 
      template specialization 'kj::Promise<void>::attach<const kj::Array<kj::ArrayPtr<const unsigned 
      char>>>' requested here
    /ekam-provider/c++header/kj/array.h:146:20: note: 'Array' has been explicitly marked deleted here
    In file included from /ekam-provider/canonical/capnp/compat/byte-stream.c++:22:
    In file included from /ekam-provider/canonical/capnp/compat/byte-stream.h:25:
    In file included from /ekam-provider/c++header/capnp/compat/byte-stream.capnp.h:9:
```

Reply:
You probably just need to mark the lambda as mutable, since right now you're trying to `mv` a const `Array`.

Reply:
lol, sigh, yup. Error message fatigue. :(